lib/stringtheory/src/lib.rs

```diff
@@ -1,58 +1,478 @@
+impl Clone for Inner {
+    fn clone(&self) -> Self {
+        let union_type = unsafe { self.discriminant.get_union_type() };
+        match union_type {
+            UnionType::Empty => Self::empty(),
+            UnionType::Owned => {
+                let owned = unsafe { self.owned };
+                let s = owned.as_str();
+
+                // We specifically try to inline here.
+                //
+                // At a high-level, when we're _given_ an owned string, we avoid inlining because we don't want to trash
+                // the underlying allocation since we may be asked to give it back later (via `MetaString::into_owned`).
+                // However, when we're _cloning_, we'd rather avoiding allocating if we can help it.
+                Self::try_inlined(s).unwrap_or_else(|| Self::owned(s.to_owned()))
+            }
+            UnionType::Static => Self {
+                _static: unsafe { self._static },
+            },
+            UnionType::Interned => {
+                let interned = unsafe { &self.interned };
+                Self::interned(interned.state.clone())
+            }
+            UnionType::Inlined => Self {
+                inlined: unsafe { self.inlined },
+            },
+        }
+    }
+}
+
+// SAFETY: None of our union variants are tied to the original thread they were created on.
+unsafe impl Send for Inner {}
+
+// SAFETY: None of our union variants use any form of interior mutability and are thus safe to be shared between
+// threads.
+unsafe impl Sync for Inner {}
+
+/// An immutable string type that abstracts over various forms of string storage.
 ///
 /// Normally, developers will work with either `String` (owned) or `&str` (borrowed) when dealing with strings. In some
 /// cases, though, it can be useful to work with strings that use alternative storage, such as those that are atomically
-/// shared (e.g. `Arc<str>`). While using those string types themselves isn't complex, using them and _also_ supporting
-/// normal string types can be complex.
+/// shared (e.g. `InternedString`). While using those string types themselves isn't complex, using them and _also_
+/// supporting normal string types can be complex.
 ///
 /// `MetaString` is an opinionated string type that abstracts over the normal string types like `String` and `&str`
-/// while also supporting alternative storage, such as interned strings from `FixedSizeInterner`.
+/// while also supporting alternative storage, such as interned strings from `FixedSizeInterner`, unifying these
+/// different variants behind a single concrete type.
 ///
 /// ## Supported types
 ///
-/// `MetaString` supports the following "modes":
+/// `MetaString` supports the following "variants":
 ///
-/// - owned (`String`)
+/// - owned (`String` and non-inlineable `&str`)
 /// - static (`&'static str`)
 /// - interned (`InternedString`)
-/// - inlined (up to 31 bytes)
+/// - inlined (up to 23 bytes)
```

Comment:
`InlinedString` didn't support 31 bytes, only 23. :)

Reply:
Yeah, this comment is old but it _used_ to be true!

Two versions ago, with the enum-based approach, we had a `Bytes` variant, which was 32 bytes, and we sized `InlinedString` to fit that.